### DISCUSSION TITLE
Sync conflicts and data loss

### DISCUSSION THREAD

**Discussion by christophberger** [remotely-save/remotely-save#76] [General]
Hi @fyears,

First of all, thanks a lot for this plugin! I have set up a test vault with Firebase (encrypted), and it works like a charm.

The only thing I am not comfortable with is automatic sync conflict resolution:

> All files and folders are compared using their local and remote "last modified time" and those with later "last modified time" wins.

What happens to the losing file? I assume it gets deleted, correct?

In this case, if the file with the earlier "last modified time" is actually the one I want to keep, this mechanism _will_ eventually cause data loss, without an option for recovering the data. And this is a quite worrisome outlook. 

I would rather want to have the losing version get renamed to something like "filename-conflict-yyyy-mm-dd-hh-mm-ss.ext". 
Thoughts?

**Comment by fyears** [comment]
> What happens to the losing file? I assume it gets deleted, correct?

yes. unless your s3 provider has some historical files saving services. I know that backblaze has some sort of this service.

> In this case, if the file with the earlier "last modified time" is actually the one I want to keep

To accomplish this, every historical version of the file is needed to be saved, because the plugin, or any software, is unable to know which version you want to keep in advance. Or any other good way?

**Comment by christophberger** [comment]
Thanks @fyears. I don't think that saving every historical version is necessary. If a conflict occurs, simply renaming the losing file instead of deleting it will do the trick. 

Example: A file on S3 has a modification time of 8am.
I modify the file on device A at 9am, and on device B at 10am. 
Then I sync both devices at 11am.

Both files are newer than the copy on S3, hence both contain changes that should not get lost.

The "latest wins" formula would delete the modifications I made on device A forever, and without notice.

If the losing file would get renamed to, say, "filename-conflict-9am-deviceA" instead of deleted, the changes from device A can survive, and I would be able to eventually find the "lost" changes and merge them if needed.

**Comment by christophberger** [comment]
Ok wait, I see the point - the two devices do not sync simultaneously, hence the situation "both files are newer than the S3 version" never occurs. Ok, that's indeed a problem.
